### PR TITLE
Normalize type of value for setter assemblers

### PR DESCRIPTION
**Pull request by WinterSilence**

### CODE REVIEW

File: src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php [hunk 1]
```diff
@@ -52,15 +57,25 @@ public function assemble(ContextInterface $context)
         try {
             $methodName = Normalizer::generatePropertyMethod('set', $property->getName());
             $class->removeMethod($methodName);
+            if (
+                $this->options->useNormalizeValue()
```
Comment (WinterSilence) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php]: `60 | ERROR | [x] Expected 0 spaces after opening bracket; newline found` pass validator in trash

File: src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php [hunk 2]
```diff
@@ -50,22 +55,25 @@ public function assemble(ContextInterface $context)
...
                 MethodGenerator::fromArray(
                     [
-                        'name' => $methodName,
+                        'name'       => $methodName,
                         'parameters' => [$parameterOptions],
                         'visibility' => MethodGenerator::VISIBILITY_PUBLIC,
-                        'body' => sprintf('$this->%1$s = $%1$s;', $property->getName()),
-                        'docblock' => DocBlockGeneratorFactory::fromArray([
+                        'body'       => sprintf($body, $property->getName(), $property->getType()),
+                        'docblock'   => DocBlockGeneratorFactory::fromArray([
```
Comment (drupol) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php:73]: Unrelated changes.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php [hunk 3]
```diff
@@ -50,22 +55,25 @@ public function assemble(ContextInterface $context)
...
                             'tags' => [
                                 [
-                                    'name' => 'param',
+                                    'name'        => 'param',
```
Comment (drupol) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php:76]: Unrelated changes.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssemblerOptions.php [hunk 4]
```diff
@@ -33,6 +38,18 @@ public function withTypeHints(bool $typeHints = true): SetterAssemblerOptions
...
         $new->typeHints = $typeHints;
+        return $new;
+    }
+
+    /**
+     * @param bool $normalizeValue
+     *
+     * @return SetterAssemblerOptions
+     */
+    public function withNormalizeValue(bool $normalizeValue = true): SetterAssemblerOptions
```
Comment (drupol) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssemblerOptions.php:49]: I would sort public method alphabetically, so this one must go under useNormalizeValue().

**Comment by WinterSilence** [comment]
@drupol done, check please

File: src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php [hunk 5]
```diff
@@ -52,15 +57,25 @@ public function assemble(ContextInterface $context)
...
+                && ! $this->options->useTypeHints()
+                && in_array($property->getType(), $this->phpTypes)
+            ) {
+                $body = '$this->%1$s = (%3$s) $%1$s;%2$sreturn $this;';
+            } else {
+                $body = '$this->%1$s = $%1$s;%2$sreturn $this;';
+            }
```
Comment (drupol) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php:68]: Why not use a simpler construction for the `$body` variable ?

```php
$cast = '';
if (
    $this->options->useNormalizeValue()
    && ! $this->options->useTypeHints()
    && in_array($property->getType(), $this->phpTypes)
) {
    $cast .= '(%3$s)';
}

$body = '$this->%1$s = ' . $cast . ' $%1$s;%2$s' . 'return $this;';
```

File: src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php [hunk 6]
```diff
@@ -50,8 +55,11 @@ public function assemble(ContextInterface $context)
         $property = $context->getProperty();
         try {
             $parameterOptions = ['name' => $property->getName()];
+            $body = '$this->%1$s = $%1$s;';
             if ($this->options->useTypeHints()) {
                 $parameterOptions['type'] = $property->getType();
+            } elseif ($this->options->useNormalizeValue() && in_array($property->getType(), $this->phpTypes)) {
+                $body = '$this->%1$s = (%2$s) $%1$s;';
```
Comment (drupol) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php:62]: We should reduce the amound of duplicated strings, just like in the example I pasted up there.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php (hunk 5 above)
Comment (WinterSilence) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php:68]: @drupol it's not simpler for me, what's about:
~~~
$body = '$this->%1$s = ';
if (
    $this->options->useNormalizeValue()
    && ! $this->options->useTypeHints()
    && in_array($property->getType(), $this->phpTypes)
) {
    $body .= '(%3$s) ';
}
$body .= '$%1$s;%2$sreturn $this;';
~~~

Comment (drupol) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php:68]: It's perfectly fine as well for sure!

**Comment by WinterSilence** [comment]
@drupol also I some confused - how be with code style in [FluentSetterAssembler.php](https://github.com/WinterSilence/soap-client/blob/master/src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php)?

**Comment by drupol** [comment]
> @drupol also I some confused - how be with code style in [FluentSetterAssembler.php](https://github.com/WinterSilence/soap-client/blob/master/src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php)?

Could you please rephrase ? I don't understand your question.

**Comment by WinterSilence** [comment]
@drupol  zend reborn as laminas, how be with class names?

**Comment by drupol** [comment]
> @drupol zend reborn as laminas, how be with class names?

yes, and ? What's your question exactly ?

**Comment by WinterSilence** [comment]
@drupol half files use other code style. FluentSetterAssembler.php as example - used extra whitespaces for arrays.

**Comment by WinterSilence** [comment]
@drupol https://github.com/WinterSilence/soap-client/blob/1351597b75e6e588b74ae8ad144d0a89834948d3/src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php#L12

File: src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php (hunk 5 above)
Comment (WinterSilence) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php:67]: @drupol fixed

**Comment by drupol** [comment]
ooh that ok I got you I think.

When contributing to an open-source project, it's mandatory that the changes you're trying to introduce are as small and easy as possible for reviewing and also well scoped.

Do not change what is not supposed to be changed, or open another pull request to change them.

Fixing other files is not in the scope of this ticket, so, don't do it.
However, you're free to open another pull request and fix them all at once.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php (hunk 6 above)
Comment (drupol) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php:62]: Same here

**Comment by WinterSilence** [comment]
@drupol  ok, then i'm done - all fixed.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssemblerOptions.php (hunk 4 above)
Comment (WinterSilence) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssemblerOptions.php:65]: @drupol done

File: src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php (hunk 5 above)
Comment (janvernieuwe) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/FluentSetterAssembler.php:67]: Maybe consider moving this to it's own `generateBody` method?

File: src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php (hunk 6 above)
Comment (janvernieuwe) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/SetterAssembler.php:62]: Same here, consider using a method.
We try not to use else (since it often points to code that needs to be structured better).

**Comment by veewee** [comment]
Closing this one. In v3, the type system will be much better.
Normalizing the type doesn't seem something we want to be doing from this package.
